apps/erlangbridge/src/gen_lsp_doc_server.erl

```diff
@@ -3,66 +3,127 @@
 -behavior(gen_server).
 -export([start_link/0]).
 
+-export([document_opened/2, document_changed/2, document_closed/1, opened_documents/0, get_document_contents/1, parse_document/1]).
+-export([project_file_added/1, project_file_changed/1, project_file_deleted/1]).
+-export([get_syntax_tree/1, get_dodged_syntax_tree/1, get_references/1]).
+-export([root_available/0, project_modules/0, get_module_file/1, get_build_dir/0, find_source_file/1]).
 -export([init/1,handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).
--export([set_document_attribute/3, remove_document/1, get_document_attribute/2, get_documents/0]).
--export([root_available/0, project_modules/0, add_project_file/1, remove_project_file/1, get_module_file/1, get_module_beam/1, get_build_dir/0]).
 
 -define(SERVER, ?MODULE).
 
--record(state, {opened, project_modules}).
+-record(state, {project_modules, files_to_parse}).
 
-start_link() ->
-    gen_server:start_link({local, ?SERVER}, ?MODULE, [],[]).
+document_opened(File, Contents) ->
+    ets:insert(document_contents, {File, Contents}).
+
+document_changed(File, Contents) ->
+    ets:insert(document_contents, {File, Contents}).
+
+document_closed(File) ->
+    ets:delete(document_contents, File).
+
+opened_documents() ->
+    [File || {File, _Contents} <- ets:tab2list(document_contents)].
+
+get_document_contents(File) ->
+    case ets:lookup(document_contents, File) of
+        [{File, Contents}] -> Contents;
+        _ -> undefined
+    end.
+
+parse_document(File) ->
+    case filename:extension(File) of
+        ".erl" ->
+            case get_document_contents(File) of
+                undefined ->
+                    io:format("Cannot find contents of document ~p~n", [File]);
```

Comment:
using gen_lsp_server:lsp_log(..) should better no ?

Reply:
Agreed on io:format. However, this is exceptional situation, since an opened document should always have contents set. So in this case instead of lsp_log I usede error_logger:error_msg.